numpy/core/einsumfunc.py

```diff
@@ -0,0 +1,597 @@
+    output_set : set
+        Set that represents the rhs side of the overall einsum subscript
+
+    Returns
+    -------
+    new_result : set
+        The indices of the resulting contraction
+    remaining : list
+        List of sets that have not been contracted
+    idx_removed : set
+        Indices removed from the entire contraction
+    idx_contraction : set
+        The indices used in the current contraction
+    """
+
+    idx_contract = set()
+    idx_remain = output_set.copy()
+    remaining = []
+    for ind, value in enumerate(input_sets):
+        if ind in positions:
+            idx_contract |= value
+        else:
+            remaining.append(value)
+            idx_remain |= value
+
+    new_result = idx_remain & idx_contract
+    idx_removed = (idx_contract - new_result)
+    remaining.append(new_result)
+    return (new_result, remaining, idx_removed, idx_contract)
+
+
+def _path_optimal(input_sets, output_set, idx_dict, memory_limit):
+    """
+    Computes all possible pair contractions, sieves the results based
+    on ``memory_limit`` and returns the lowest cost path.
+
+    Paramaters
+    ----------
+    input_sets : list
+        List of sets that represent the lhs side of the einsum subscript
+    output_set : set
+        Set that represents the rhs side of the overall einsum subscript
+    idx_dict : dictionary
+        Dictionary of index sizes
+    memory_limit : int
+        The maximum number of elements in a temporary array.
+
+    Returns
+    -------
+    path : list
+        The optimal order of pair contractions.
+    """
+
+    current = [(0, [], input_sets)]
+    for iteration in range(len(input_sets) - 1):
+        new = []
+
+        # Grab all unique pairs
+        comb_iter = []
+        for x in range(len(input_sets) - iteration):
+            for y in range(x + 1, len(input_sets) - iteration):
+                comb_iter.append((x, y))
+
+        for curr in current:
+            cost, positions, remaining = curr
+            for con in comb_iter:
+
+                contract = _find_contraction(con, remaining, output_set)
+                new_result, new_input_sets, idx_removed, idx_contract = contract
+
+                # Sieve the results based on memory_limit
+                new_size = _compute_size_by_dict(new_result, idx_dict)
+                if new_size > memory_limit:
+                    continue
+
+                # Find cost
+                new_cost = _compute_size_by_dict(idx_contract, idx_dict)
+                if len(idx_removed) > 0:
+                    new_cost *= 2
+
+                # Build (total_cost, positions, indices_remaining)
+                new_cost += cost
+                new_pos = positions + [con]
+                new.append((new_cost, new_pos, new_input_sets))
+
+        # Update list to iterate over
+        current = new
+
+    # If we have not found anything return single einsum contraction
+    if len(new) == 0:
```

Comment:
I can add another conditional here. The `einsum` function is constructed in a way so that this cannot happen. As a general question should internal functions be fully protected in this way?